src/smartHandler.ts

```diff
@@ -97,26 +102,87 @@ export class SMARTHandler implements Authorization {
                 `User identity found does not conform to the expected format: ${SMARTHandler.FHIR_USER_REGEX}`,
             );
             throw new UnauthorizedError("Requester's identity is in the incorrect format");
+        } else if (request.bulkDataAuth) {
+            const fhirUser = this.getFhirUser(response.data);
+            if (fhirUser.hostname !== this.apiUrl || !this.adminAccessTypes.includes(fhirUser.resourceType)) {
+                throw new UnauthorizedError('User does not have permission for requested operation');
+            }
+        }
+
+        const userIdentity = clone(response.data);
+        userIdentity.scopes = scopes;
+        return userIdentity;
+    }
+
+    private getScopes(decoded: Record<string, any>): string[] {
+        if (this.config.scopeValueType === 'space' && typeof decoded[this.config.scopeKey] === 'string') {
+            return decoded[this.config.scopeKey].split(' ');
+        }
+        if (this.config.scopeValueType === 'array' && Array.isArray(decoded[this.config.scopeKey])) {
+            return decoded[this.config.scopeKey];
         }
-        return response.data;
+        return [];
     }
 
     // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars
     async isAccessBulkDataJobAllowed(request: AccessBulkDataJobRequest): Promise<void> {
-        // TODO not supported for now
-        throw new UnauthorizedError('Bulk access is not allowed');
+        if (request.userIdentity.sub !== request.jobOwnerId) {
+            throw new UnauthorizedError('User does not have permission to access this Bulk Data Export job');
+        }
     }
 
-    // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars
     async isBundleRequestAuthorized(request: AuthorizationBundleRequest): Promise<void> {
-        // TODO not supported for now
-        throw new UnauthorizedError('Bundle operation is not authorzied');
+        const { scopes } = request.userIdentity;
+        request.requests.forEach((req: BatchReadWriteRequest) => {
+            if (!this.areScopesSufficient(scopes, req.operation, req.resourceType)) {
+                console.error(
+                    `User supplied scopes are insufficient\nscopes: ${scopes}\noperation: ${req.operation}\nresourceType: ${req.resourceType}`,
+                );
+                throw new UnauthorizedError('Bundle operation is not authorized');

```

Comment:
Change it to be more specific
```suggestion
                throw new UnauthorizedError('An operation within the bundle is not authorized');

```

Reply:
updated